File: kinotic-js/workspace/packages/system-api/src/api/model/workload/WorkloadReservation.ts
/**
 * The room one workload holds on a VmNode: its CPU, memory and disk while its VM runs, and its
 * disk alone once the run has ended, since a VM that is kept for a restart keeps its disk on the
 * node. A workload holds at most one reservation per node, keyed by its id.
 */
export interface WorkloadReservation {
    /** The id of the workload holding the reservation. */
    workloadId: string
    /** CPU the workload's VM is allotted, in cores; a fraction is a share of one core. */
    cpus: number
    /** Memory the workload's VM is allotted, in megabytes. */
    memoryMb: number
    /** Disk the workload's VM is allotted, in megabytes. */
    diskMb: number
    /**
     * True while the workload's run holds its CPU and memory; false once the run has ended and
     * only the disk is still held.
     */
    running: boolean
}

File: kinotic-system-api/src/main/java/org/kinotic/system/api/model/workload/WorkloadReservation.java
package org.kinotic.system.api.model.workload;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.Accessors;
import org.kinotic.management.api.model.workload.Workload;

/**
 * The room one workload's run holds on a {@link VmNode}: the CPU, memory and disk its VM is sized
 * for, from the moment it is placed until its run ends or it is destroyed. A workload holds at most
 * one reservation per node, keyed by its id, so reserving or releasing the same workload twice
 * changes nothing the second time.
 */
@Getter
@Setter
@Accessors(chain = true)
@NoArgsConstructor
public class WorkloadReservation {

    /**
     * The id of the workload holding the reservation.
     */
    private String workloadId;

    /**
     * CPU the workload's VM is allotted, in cores; a fraction is a share of one core.
     */
    private double cpus;

    /**
     * Memory the workload's VM is allotted, in megabytes.
     */
    private int memoryMb;

    /**
     * Disk the workload's VM is allotted, in megabytes.
     */
    private int diskMb;

    /**
     * The reservation a workload's run needs: everything the workload is sized for.
     */
    public static WorkloadReservation forRun(Workload workload) {
        return new WorkloadReservation().setWorkloadId(workload.getId())
                                        .setCpus(workload.getCpus())
                                        .setMemoryMb(workload.getMemoryMb())
                                        .setDiskMb(workload.getDiskSizeMb());
    }
}
